### PR TITLE
Add migration plan for monomorphic Data.List

### PR DESCRIPTION
**Pull request by Bodigrim**
As discussed in #22.

### CODE REVIEW

File: guides/monomorphic-data-list.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# What
+
+Back in 2019 [CLC has approved](https://gitlab.haskell.org/ghc/ghc/-/issues/20025#note_363426) monomorphisation of `Data.List`.
```
Comment (konsumlamm) [inline guides/monomorphic-data-list.md:3]:
```suggestion
Back in 2019, the [CLC has approved](https://gitlab.haskell.org/ghc/ghc/-/issues/20025#note_363426) the monomorphisation of `Data.List`.
```

File: guides/monomorphic-data-list.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+At the moment many functions from [`Data.List`](https://hackage.haskell.org/package/base-4.16.1.0/docs/Data-List.html) are actually polymorphic by container, allowing any `Foldable f`. The approved proposal was to monomorphise such functions to work on lists only, essentially returning to [`GHC.OldList`](https://hackage.haskell.org/package/base-4.16.1.0/docs/GHC-OldList.html) signatures, predating Foldable-Traversable Proposal. The justification for the change is available [here]. Note that this change does not affect `Prelude`, which will continue to export entities polymorphic by container.
```
Comment (konsumlamm) [inline guides/monomorphic-data-list.md:5]:
```suggestion
At the moment, many functions from [`Data.List`](https://hackage.haskell.org/package/base-4.16.1.0/docs/Data-List.html) are actually polymorphic by container, allowing any `Foldable f`. The approved proposal was to monomorphise such functions to work on lists only, essentially returning to [`GHC.OldList`](https://hackage.haskell.org/package/base-4.16.1.0/docs/GHC-OldList.html) signatures, predating the Foldable-Traversable Proposal. The justification for the change is available [here]. Note that this change does not affect `Prelude`, which will continue to export entities polymorphic by container.
```

Comment (konsumlamm) [inline guides/monomorphic-data-list.md:5]: The link for `[here]` is missing.

File: guides/monomorphic-data-list.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+* Uniformity with other container modules (e. g., `Data.Set`, `Data.Map`).
+* This makes Haskell easier to learn. Haskell learners typically start with lists, having to understand typeclasses / `Foldable` before then doesn't make any sense, is unnecessary noise to the learner, and is confusing during reading of the documentation and in error messages. I have heard several arguments from teachers of Haskell that students find the `Foldable`-generic interface of `Data.List` to be more confusing than `GHC.OldList`, for example.
+* Generalised methods rely on inlining for dictionary elimination. Monomorphic ones do not.
+* Monormophisation allows for an easier expansion of the module.
+
+# When
+
+The change was approved without a specific migration plan. The current CLC [developed](https://github.com/haskell/core-libraries-committee/issues/22#issuecomment-1167711639) the following approach:
+
+1. Wait until GHC X.Y makes [`-Wcompat-unqualified-imports`](http://downloads.haskell.org/ghc/9.2.2/docs/html/users_guide/using-warnings.html?highlight=wcompat#ghc-flag--Wcompat-unqualified-imports) a part of `-Wall`, either directly or by including all `-Wcompat` into `-Wall`.
```
Comment (konsumlamm) [inline guides/monomorphic-data-list.md:22]:
```suggestion
1. Wait until GHC X.Y makes [`-Wcompat-unqualified-imports`](http://downloads.haskell.org/ghc/9.2.2/docs/html/users_guide/using-warnings.html?highlight=wcompat#ghc-flag--Wcompat-unqualified-imports) a part of `-Wall`, either directly or by including all of `-Wcompat` into `-Wall`.
```

File: guides/monomorphic-data-list.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# What
+
+Back in 2019, the [CLC has approved](https://gitlab.haskell.org/ghc/ghc/-/issues/20025#note_363426) the monomorphisation of `Data.List`.
```
Comment (ulysses4ever) [inline guides/monomorphic-data-list.md:3]:
```suggestion
Back in 2019, the [CLC approved](https://gitlab.haskell.org/ghc/ghc/-/issues/20025#note_363426) the monomorphisation of `Data.List`.
```
(From what I understand about English grammar, pointing to an event in the past requires past simple, but feel free to disregard, of course.)

**Comment by Bodigrim** [comment]
Rebased and squashed.

**Pull request merged**